pkg/component/networking/nodelocaldns/nodelocaldns.go

```diff
@@ -158,368 +190,181 @@ func (n *nodeLocalDNS) Deploy(ctx context.Context) error {
+		}
+	} else {
+		managedResourceExists = true
+	}
+
+	// Check if at least one Kubernetes node has the cleanup label
+	nodeList := &corev1.NodeList{}
+	if err := n.values.ShootClient.List(ctx, nodeList, client.MatchingLabels{
+		labelKeyCleanupRequired: "true",
+	}); err != nil {
+		return fmt.Errorf("failed to list nodes with cleanup label: %w", err)
+	}
+
+	// If the managed resource does not exist and no nodes have the cleanup label, return early no cleanup needed
+	if !managedResourceExists && len(nodeList.Items) == 0 {
+		return nil
+	}
+
+	// Mark nodes for cleanup if the managed resource exists
+	if managedResourceExists {
+		if err := n.markNodeForCleanup(ctx); err != nil {
+			return fmt.Errorf("failed to mark nodes for cleanup: %w", err)
+		}
+	}
+
+	// Delete resources
 	if err := kubernetesutils.DeleteObjects(ctx, n.client,
 		n.emptyScrapeConfig(""),
 		n.emptyScrapeConfig("-errors"),
 	); err != nil {
 		return err
 	}
 
-	return managedresources.DeleteForShoot(ctx, n.client, n.namespace, managedResourceName)
-}
+	if err := managedresources.DeleteForShoot(ctx, n.client, n.namespace, managedResourceName); err != nil {
+		return err
+	}
 
-// TimeoutWaitForManagedResource is the timeout used while waiting for the ManagedResources to become healthy
-// or deleted.
-var TimeoutWaitForManagedResource = 2 * time.Minute
+	if v1beta1helper.IsKubeProxyIPVSMode(n.values.KubeProxyConfig) {
+		return nil
+	}
 
-func (n *nodeLocalDNS) Wait(ctx context.Context) error {
+	// Wait until the managed resource is successfully deleted and go through node list again to add cleanup label
 	timeoutCtx, cancel := context.WithTimeout(ctx, TimeoutWaitForManagedResource)
 	defer cancel()
+	if err := managedresources.WaitUntilDeleted(timeoutCtx, n.client, n.namespace, managedResourceName); err != nil {
+		return err
+	}
 
-	return managedresources.WaitUntilHealthy(timeoutCtx, n.client, n.namespace, managedResourceName)
-}
+	nodeList = &corev1.NodeList{}
+	if err := n.values.ShootClient.List(ctx, nodeList); err != nil {
+		return fmt.Errorf("failed to list all nodes: %w", err)
+	}
 
-func (n *nodeLocalDNS) WaitCleanup(ctx context.Context) error {
-	timeoutCtx, cancel := context.WithTimeout(ctx, TimeoutWaitForManagedResource)
-	defer cancel()
+	var taskFns []flow.TaskFn
+	if managedResourceExists {
+		for _, node := range nodeList.Items {
+			// Skip nodes that already have the label
+			if val, ok := node.Labels[labelKeyCleanupRequired]; ok && val == "true" {
+				continue
+			}
+
+			node := node // capture range variable
+			taskFns = append(taskFns, func(ctx context.Context) error {
+				patch := client.MergeFrom(node.DeepCopy())
+				metav1.SetMetaDataLabel(&node.ObjectMeta, labelKeyCleanupRequired, "true")
+				if err := n.values.ShootClient.Patch(ctx, &node, patch); err != nil {
+					return fmt.Errorf("failed to add cleanup label to recently joined node %s: %w", node.Name, err)
+				}
+				return nil
+			})
+		}
```

Comment:
Can't we just call the `markNodeForCleanup` func again?

Reply:
Maybe then you could create a new function for labeling the nodes and call it in both places.